director/jobs/api/views.py

```diff
@@ -0,0 +1,487 @@
+from django.utils import timezone
+from drf_yasg.utils import swagger_auto_schema
+from rest_framework import mixins, permissions, status, viewsets
+from rest_framework.exceptions import PermissionDenied
+from rest_framework.decorators import action
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from accounts.views import Account, AccountPermissionsMixin, AccountPermissionType
+from jobs.models import Job, Worker, WorkerStatus, Zone
+from jobs.api.serializers import (
+    JobListSerializer,
+    JobCreateSerializer,
+    JobRetrieveSerializer,
+    JobUpdateSerializer,
+    WorkerSerializer,
+    WorkerStatusSerializer,
+    ZoneSerializer,
+    ZoneCreateSerializer,
+)
+from jobs import jobs
+
+logger = logging.getLogger(__name__)
+
+
+class JobsViewSet(
+    mixins.ListModelMixin,
+    mixins.CreateModelMixin,
+    mixins.RetrieveModelMixin,
+    mixins.UpdateModelMixin,
+    viewsets.GenericViewSet,
+):
+
+    # Configuration
+
+    def get_permissions(self):
+        """
+        Get the list of permissions that the current action requires.
+
+        The `partial_update` action requires a staff user (an internal bot),
+        others just require authentication.
+        """
+        return (
+            [permissions.IsAdminUser()]
+            if self.action == "partial_update"
+            else [permissions.IsAuthenticated()]
+        )
+
+    def get_queryset(self):
+        """
+        Override of `GenericAPIView.get_queryset`.
+
+        Returns the list of jobs that the user has view access to.
+        """
+        # TODO: Also include projects where the user has view access
+        return Job.objects.filter(creator=self.request.user)
+
+    def check_object_permissions(self, request: Request, job: Job):
+        """
+        Override of `APIView.check_object_permissions`.
+
+        Checks that the user is either the creator of the job,
+        or has view access to the project.
+        """
+        # TODO: Also permit users with view access to the project
+        if request.user != job.creator:
+            raise PermissionDenied
+
+    def get_serializer_class(self):
+        """
+        Override of `GenericAPIView.get_serializer_class`.
+
+        Returns different serializers for different views.
+        """
+        try:
+            return {
+                "list": JobListSerializer,
+                "create": JobCreateSerializer,
+                "execute": JobCreateSerializer,
+                "retrieve": JobRetrieveSerializer,
+                "update": JobUpdateSerializer,
+                "partial_update": JobUpdateSerializer,
+                "cancel": JobRetrieveSerializer,
+                "connect": JobRetrieveSerializer,
+            }[self.action]
+        except KeyError:
+            logger.error("No serializer defined for action {}".format(self.action))
+            return JobRetrieveSerializer
+
+    # Standard views
+
+    def list(self, request: Request):
```

Comment:
This and `create` below are redundant as they just override the super method without adding anything?